integration/distributor_test.go

```diff
@@ -1289,3 +1289,200 @@ func assertStats(t *testing.T, expectedStats promRemote.WriteResponseStats, res
 	require.NoError(t, err, "exemplars stats header value should be an integer")
 	require.Equal(t, expectedStats.Exemplars, exemplars, "wrong exemplars stats header value")
 }
+
+func TestDistributor_NameValidation(t *testing.T) {
+	for _, scheme := range []model.ValidationScheme{model.LegacyValidation, model.UTF8Validation} {
+		t.Run(scheme.String(), func(t *testing.T) {
+			testDistributorNameValidation(t, scheme)
+		})
+	}
+}
+
+func testDistributorNameValidation(t *testing.T, validationScheme model.ValidationScheme) {
+	s, err := e2e.NewScenario(networkName)
+	require.NoError(t, err)
+	defer s.Close()
+
+	require.NoError(t, writeFileToSharedDir(s, "runtime.yaml", []byte("")))
+
+	consul := e2edb.NewConsul()
+	minio := e2edb.NewMinio(9000, blocksBucketName)
+	require.NoError(t, s.StartAndWaitReady(consul, minio))
+
+	baseFlags := map[string]string{
+		"-distributor.ingestion-tenant-shard-size":           "0",
+		"-ingester.ring.heartbeat-period":                    "1s",
+		"-distributor.ha-tracker.enable":                     "true",
+		"-distributor.ha-tracker.enable-for-all-users":       "true",
+		"-distributor.ha-tracker.store":                      "consul",
+		"-distributor.ha-tracker.consul.hostname":            consul.NetworkHTTPEndpoint(),
+		"-distributor.ha-tracker.prefix":                     "prom_ha/",
+		"-validation.name-validation-scheme":                 validationScheme.String(),
+		"-timeseries-unmarshal-caching-optimization-enabled": "false",
+	}
+
+	flags := mergeFlags(
+		BlocksStorageFlags(),
+		BlocksStorageS3Flags(),
+		baseFlags,
+	)
+
+	// We want only the distributor to be reloading runtime config.
+	distributorFlags := mergeFlags(flags, map[string]string{
+		"-runtime-config.file":          filepath.Join(e2e.ContainerSharedDir, "runtime.yaml"),
+		"-runtime-config.reload-period": "100ms",
+		// Set non-zero default for number of exemplars. That way our values used in the test (0 and 100) will show up in runtime config diff.
+		"-ingester.max-global-exemplars-per-user": "3",
+	})
+
+	// Ingester will not reload runtime config.
+	ingesterFlags := mergeFlags(flags, map[string]string{
+		// Ingester will always see exemplars enabled. We do this to avoid waiting for ingester to apply new setting to TSDB.
+		"-ingester.max-global-exemplars-per-user": "100",
+	})
+
+	// Start Mimir components.
+	distributor := e2emimir.NewDistributor("distributor", consul.NetworkHTTPEndpoint(), distributorFlags)
+	ingester := e2emimir.NewIngester("ingester", consul.NetworkHTTPEndpoint(), ingesterFlags)
+	querier := e2emimir.NewQuerier("querier", consul.NetworkHTTPEndpoint(), flags)
```

Comment:
we don't use this querier in the test. Can we actually query these metrics back? Right now this looks like a validation unit test. I think we should use the overhead of an integration test to run more assertions. I'd even go as far as spinning up a query-frontend

Reply:
Good suggestion, I adjusted the test!